File: Hihocoder/1309.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <queue>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <cstring>
#include <bitset>
#include <algorithm>

using namespace std;

const int N = 100100;
struct Interval {
	int start, end;
	bool operator< (const Interval& a) const {
		return end > a.end;
	}
} interval[N];

priority_queue<Interval> pq;
bool compare(const Interval &a, const Interval &b) {
	return a.start < b.start;
}

int main() {
	// freopen("input", "r", stdin);
	int n; cin >> n;
	for (int i = 0; i < n; i++) {
		int start, end; cin >> start >> end;
		interval[i].start = start;
		interval[i].end = end;
	}
	sort(interval, interval + n, compare);
	int ret = 0;
	for (int i = 0; i < n; i++) {
		if (!pq.empty() && pq.top().end <= interval[i].start) {
			Interval top = pq.top();
			pq.pop();
			top.end = interval[i].end;
			pq.push(top);
		} else {
			ret++;
			pq.push(interval[i]);
		}
	}
	cout << ret << endl;
	return 0;
}

File: Hihocoder/1309.java
import java.io.*;
import java.math.BigInteger;
import java.util.*;

public class Main {

    public static void main(String[] args) throws FileNotFoundException {
//        Scanner input = new Scanner(new File("test.txt"));
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        Interval[] interval = new Interval[n];
        for(int i = 0; i < n; i++) {
            interval[i] = new Interval(input.nextInt(), input.nextInt());
        }
        Arrays.sort(interval, new Comparator<Interval>(){
            public int compare(Interval i1, Interval i2) {
                if(i1.start == i2.start)
                    return i1.end - i2.end;
                else
                    return i1.start - i2.start;
            }
        });
        PriorityQueue<Integer> end = new PriorityQueue<Integer>();
        int ret = 0;
        for(int i = 0; i < n; i++) {
            Interval cur = interval[i];
            if(i == 0) {
                ret++;
                end.add(cur.end);
            }
            else {
                if(cur.start >= end.peek()) {
                    end.poll();
                }
                else
                    ret++;
                end.add(cur.end);
            }
        }
        System.out.println(ret);
    }
}
class Interval {
    int start, end;
    Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
}